### PR TITLE
feat: introduce company and usercompany structure

### PR DESCRIPTION
**Pull request by rebelchris**
Adds the new structure for company and user_company tables.
Also added all necessary mutations and query changes.

## Todo:
- [x] test cases

You should be able to review this part already :)

### CODE REVIEW

File: src/data-source.ts [hunk 1]
```diff
@@ -16,7 +16,7 @@ export const AppDataSource = new DataSource({
     max: 20,
     idleTimeoutMillis: 120000,
   },
-  logging: false,
+  logging: true,
```
Comment (omBratteng) [inline src/data-source.ts]: Accident?

File: src/schema/users.ts [hunk 2]
```diff
@@ -1320,6 +1377,22 @@ export const resolvers: IResolvers<unknown, BaseContext> = traceResolvers<
...
+        builder.queryBuilder = builder.queryBuilder
+          .andWhere(`${builder.alias}."userId" = :userId`, {
+            userId: ctx.userId,
+          })
+          .andWhere(`${builder.alias}."verified" = true`);
+
+        return builder;
+      });
+      if (userCompanies.length === 0) {
+        throw new NotFoundError('No companies found');
```
Comment (omBratteng) [inline src/schema/users.ts:1391]: Do we need to throw error, or just return empty array or null?

File: src/schema/users.ts [hunk 3]
```diff
@@ -1636,6 +1709,75 @@ export const resolvers: IResolvers<unknown, BaseContext> = traceResolvers<
 
       return { _: true };
     },
+    addUserCompany: async (
```
Comment (omBratteng) [inline src/schema/users.ts:1712]: Does this check if the email already exist?

File: src/schema/users.ts [hunk 4]
```diff
@@ -1320,6 +1377,22 @@ export const resolvers: IResolvers<unknown, BaseContext> = traceResolvers<
 
       return personalizedDigest;
     },
+    companies: async (_, __, ctx: AuthContext, info) => {
```
Comment (rebelchris) [inline src/schema/users.ts:1380]: Missing return type here..

File: src/schema/users.ts (hunk 2 above)
Comment (rebelchris) [reply src/schema/users.ts]: Ah good point, empty array is fine :)
adjusted (Also are you not on leave?)

File: src/schema/users.ts (hunk 3 above)
Comment (rebelchris) [reply src/schema/users.ts:1712]: Ah good one, will need to add a test for it, it's a unique index so think the query will already fail on it.

File: src/schema/users.ts (hunk 2 above)
Comment (omBratteng) [reply src/schema/users.ts]: > (Also are you not on leave?)
🙈

File: src/schema/users.ts (hunk 3 above)
Comment (omBratteng) [reply src/schema/users.ts:1712]: If the email exists, but claimed by same user and not verified, we should just "ignore" that it exists and just update verification code
In the case when the user stops midway in the verification process (i.e. computer dies or something) and before the cleanup job gets it. [Thread](https://dailydotdev.slack.com/archives/C07EA1E1XME/p1723730384460159?thread_ts=1723727031.217919&cid=C07EA1E1XME)

File: src/entity/UserCompany.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+import { User } from './user';
+import { Company } from './Company';
+
+@Entity()
+export class UserCompany {
+  @Column({ default: () => 'now()' })
+  createdAt: Date;
+
+  @Column({ default: () => 'now()' })
+  updatedAt: Date;
```
Comment (capJavert) [inline src/entity/UserCompany.ts:19]: We can do `CreateDateColumn` i `UpdateDateColumn`

File: src/entity/UserCompany.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+  @Column()
+  companyId: string;
+
+  @OneToOne(() => Company, {
+    nullable: true,
+    lazy: true,
+    onDelete: 'CASCADE',
+  })
+  @JoinColumn({ name: 'companyId' })
```
Comment (capJavert) [inline src/entity/UserCompany.ts:50]: Typeorm does not know its `companyId` automatically?

File: src/entity/UserCompany.ts [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+  @Column({ default: false })
+  verified: boolean;
+
+  @PrimaryColumn({ type: 'text' })
+  @Index({ unique: true })
+  email: string;
+
+  @Column({ type: 'text' })
+  @Index({ unique: true })
+  code: string;
```
Comment (capJavert) [inline src/entity/UserCompany.ts:30]: Why does code need to be unique?

**Review by capJavert — COMMENTED** [review]
Nice work. Some comments/questions

File: src/schema/users.ts (hunk 3 above)
Comment (rebelchris) [reply src/schema/users.ts:1712]: So resubmitting email will just generate a new code?
I'll add that :)

File: src/entity/UserCompany.ts (hunk 5 above)
Comment (rebelchris) [reply src/entity/UserCompany.ts:19]: What difference do they actually make under the hood?
(I will modify them)

File: src/entity/UserCompany.ts (hunk 6 above)
Comment (rebelchris) [reply src/entity/UserCompany.ts:50]: Missed it before I added the column definition:
```
@Column()
  companyId: string;
```

File: src/entity/UserCompany.ts (hunk 7 above)
Comment (rebelchris) [reply src/entity/UserCompany.ts:30]: Hmm yeah actually doesn't have to be :)

File: src/schema/users.ts [hunk 8]
```diff
@@ -1636,6 +1709,75 @@ export const resolvers: IResolvers<unknown, BaseContext> = traceResolvers<
...
       return { _: true };
     },
+    addUserCompany: async (
+      _,
+      { email }: { email: string },
+      ctx: AuthContext,
+    ): Promise<GQLEmptyResponse> => {
+      if (!email?.length) {
+        throw new ValidationError('Email is required');
+      }
```
[the remark this answers is not in the document]
Comment (rebelchris) [reply src/schema/users.ts:1719]: I'll check with tests :)

File: src/entity/UserCompany.ts (hunk 5 above)
Comment (capJavert) [reply src/entity/UserCompany.ts:19]: `UpdateDateColumn` should update when you do `save` on entity.

File: src/schema/users.ts [hunk 9]
```diff
@@ -1636,6 +1711,79 @@ export const resolvers: IResolvers<unknown, BaseContext> = traceResolvers<
...
+          email,
+        });
+
+      if (existingUserCompanyEmail) {
+        if (existingUserCompanyEmail.userId !== ctx.userId) {
+          throw new ValidationError('Email already exists');
+        }
+
+        const updatedRecord = { ...existingUserCompanyEmail, code };
+        await ctx.con.getRepository(UserCompany).save(updatedRecord);
```
Comment (omBratteng) [inline src/schema/users.ts:1740]: IIRC if we use upsert, it'll update if exists, or insert if not?

**Review by omBratteng — APPROVED** [review]
Looks good, haven't looked at tests, too much for mobile 🙈

File: src/schema/users.ts (hunk 9 above)
Comment (rebelchris) [reply src/schema/users.ts:1740]: Hmm not sure on typeorm upsert, I vaguely recall it having some issues.
CC: @capJavert you recall anything for upsert?

(We still need the split query anyway for fetching as we don't want to update someone else's entry)

Comment (capJavert) [reply src/schema/users.ts:1740]: save should also do the same but upsert allows providing on conflict part

File: src/entity/Company.ts [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+@Entity()
+export class Company {
+  @PrimaryColumn({ type: 'text' })
+  id: string;
+
+  @Column({ type: 'text' })
+  name: string;
+
+  @CreateDateColumn()
+  createdAt: Date;
```
Comment (capJavert) [inline src/entity/Company.ts:12]: Just for the sake of it, do we want updatedAt here as well, useful for most entities?

File: src/schema/users.ts [hunk 11]
```diff
@@ -1646,6 +1721,79 @@ export const resolvers: IResolvers<unknown, BaseContext> = traceResolvers<
...
+      _,
+      { email }: { email: string },
+      ctx: AuthContext,
+      info,
+    ): Promise<GQLUserCompany[]> => {
+      await ctx.con
+        .getRepository(UserCompany)
+        .delete({ email, userId: ctx.userId });
+
+      return getUserCompanies(_, ctx, info);
```
Comment (capJavert) [inline src/schema/users.ts:1772]: Seems wasteful to fetch all companies on delete? Why do you need it?

File: src/entity/Company.ts (hunk 10 above)
Comment (rebelchris) [reply src/entity/Company.ts:12]: We could add it yeah, don't see any harm in it.

File: src/schema/users.ts [hunk 12]
```diff
@@ -727,6 +764,11 @@ export const typeDefs = /* GraphQL */ `
     Get integrations for the user
     """
     userIntegrations: UserIntegrationConnection @auth
+
+    """
+    Get companies for user
+    """
+    companies: [UserCompany] @auth
```
Comment (capJavert) [inline src/schema/users.ts:771]: Just noticed this allows querying email for other users over `companies` field? Also do we want to limit in customRelation to something like 10 or 50 so we don't return hundreds of rows if someone tries to be naughty

File: src/graphorm/index.ts [hunk 13]
```diff
@@ -119,6 +120,38 @@ const obj = new GraphORM({
...
+          isMany: true,
+          customRelation: (ctx, parentAlias, childAlias, qb): QueryBuilder =>
+            qb
+              .innerJoin(
+                UserCompany,
+                'uc',
+                `"${childAlias}"."id" = uc."companyId"`,
+              )
+              .where('uc.verified = true')
+              .andWhere(`uc."userId" = "${parentAlias}".id`),
```
Comment (capJavert) [inline src/graphorm/index.ts:134]:
```suggestion
              .andWhere(`uc."userId" = "${parentAlias}".id`)
              .limit(50),
```

just so we don't return hundreds of rows if somebody tries to be naughty, we do the same for `privilegedMembers`

**Review by capJavert — APPROVED** [review]
Awesome work, just last two comments to check.

File: src/graphorm/index.ts (hunk 13 above)
Comment (rebelchris) [reply src/graphorm/index.ts:134]: For now you can actually only add 1 anyway, we dont support more from fe.

We can tackle this maybe in follow up once we ever do?

Comment (capJavert) [reply src/graphorm/index.ts:134]: But from API directly someone could do it 😆

Comment (capJavert) [reply src/graphorm/index.ts:134]: Someone naughty xD

**Pull request merged**